### ISSUE TITLE
This tool doesn't work on M1 (aarch64 architecture) Macs.

### ISSUE THREAD

**Issue by jameslittle230** [bheisler/cargo-criterion#21]
Hi! A non-blocker, but I've run into a snag where I have some Criterion tests, and the `cargo criterion` command fails on my M1 Mac but not on an Ubuntu VM.

Rust's support for M1 macs is laid out in https://github.com/rust-lang/rust/issues/73908

Thanks for building criterion! It's an invaluable tool.

**Comment by bheisler** [comment]
Hey, thanks for the bug report!

I don't have an M1 Mac, or any Mac, nor would I have the first idea of how to develop on one if I did. I probably can't fix this one myself. I'd be open to pull requests though.

**Comment by jameslittle230** [comment]
Makes total sense, thanks for responding :)

For tracking purposes, here's the `RUST_BACKTRACE` that I see when I run `$ cargo criterion`:

```
Error: Unable to open connection to bench target basic

Caused by:
    Resource temporarily unavailable (os error 35)
Gnuplot not found, using plotters backend

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: IoError(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" })', /Users/james/.cargo/registry/src/github.com-1ecc6299db9ec823/criterion-0.3.3/src/lib.rs:1438:18
stack backtrace:
   0: rust_begin_unwind
             at /rustc/14b27555a03f4abf9b8d80ea34f59144e2ded5e4/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/14b27555a03f4abf9b8d80ea34f59144e2ded5e4/library/core/src/panicking.rs:92:14
   2: core::option::expect_none_failed
             at /rustc/14b27555a03f4abf9b8d80ea34f59144e2ded5e4/library/core/src/option.rs:1300:5
   3: criterion::Criterion<M>::benchmark_group
   4: basic::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by ubamrein** [comment]
Actually I get the same issue on my standard Intel-Macbook. I could trace it down to the "set_nonblocking" line  here:
https://github.com/bheisler/cargo-criterion/blob/2333384d963c88fffe2e37c7345a35c377ee61cf/src/bench_target.rs#L40 

As a minimal reproducing example (this fails when trying to read, so e.g test with `nc localhost <port>`):

```rust
use std::{
    io::{Read, Write},
    net::TcpListener,
};

fn main() {
    let listener = TcpListener::bind("localhost:0").unwrap();
    listener.set_nonblocking(true);
    let addr = listener.local_addr().unwrap();
    let port = addr.port();
    println!("{}", port);
    loop {
        match listener.accept() {
            Ok((mut socket, addr)) => {
                socket.write_all(b"Hallo").unwrap();
                let mut test_buf = [0u8; 3];
                socket.read_exact(&mut test_buf).unwrap();
                println!("{:?}", String::from_utf8_lossy(&test_buf));
            }
            Err(_) => {}
        }
    }
}

``` 


Based on the documentation of ['TcpStream'](https://doc.rust-lang.org/std/net/struct.TcpStream.html), when using 'non_blocking' sockets, we need to check the I/O for 'std::io::ErrorKind::WouldBlock'. Extending the example before:

(this works)
```rust
use std::{
    io::{Read, Write},
    net::TcpListener,
};

fn main() {
    let listener = TcpListener::bind("localhost:0").unwrap();
    listener.set_nonblocking(true);
    let addr = listener.local_addr().unwrap();
    let port = addr.port();
    println!("{}", port);
    loop {
        match listener.accept() {
            Ok((mut socket, addr)) => {
                socket.write_all(b"Hallo").unwrap();
                let mut test_buf = [0u8; 3];
                loop {
                    match socket.read_exact(&mut test_buf) {
                        Ok(_) => break,
                        Err(ref e) if e.kind() == std::io::ErrorKind::WouldBlock => {
                            std::thread::sleep(std::time::Duration::from_millis(10));
                        }
                        Err(e) => panic!("{:?}", e)
                    }
                    ;
                }
                println!("{:?}", String::from_utf8_lossy(&test_buf));
            }
            Err(_) => {}
        }
    }
}

```

So, not really sure what the reason behind the 'nonblocking' is, but I think in this case it would be fine to not set it? 

Based on the response, I could try preparing a PR :) 

If the 'nonblocking' is not really needed, I would just set it to false and test it locally. Otherwise I would need to lookup the respective 'polling' functions (e.g. epoll) to make it work on all platforms.

**Comment by inflation** [comment]
Should be the same issue of #12

**Comment by Luapulu** [comment]
Just wanted to check in to say I have the same issue on my Intel Mac.

**Comment by Lemmih** [comment]
Fixed in `cargo-criterion-1.1.0`.
